File: agent-driven-architecture-main/examples/kotlin/block/analysis/src/main/kotlin/adr/blocks/analysis/Slice.kt
// ── blocks/analysis/slice — the tier's own state ───────────────────────────
// Copy-on-write; never mutate the input. Structural equality is what makes the
// replay comparison meaningful (14.1) — and here it is doing real work: the note
// stores the WHOLE sealed `Recall`, so a re-folded timeline can be compared variant
// for variant against the live one.

package adr.blocks.analysis

import adr.spine.pure.Recall
import adr.spine.pure.Timestamp

/**
 * One recall, as it was actually resolved. The variant is kept, not flattened to
 * text: `Fresh("x")` and `LastKnown("x")` are different facts about the same string,
 * and an audit that cannot tell them apart cannot answer "was the agent reasoning
 * over stale input?".
 */
data class AnalysisNote(val at: Timestamp, val recall: Recall)

data class AnalysisSlice(
    val notes: List<AnalysisNote> = emptyList(),
    /** What THIS tier published. A fast tier never has any; a deep tier's is its output. */
    val published: List<String> = emptyList(),
) {
    // No companion: a companion member has no instance, the same defect as a top-level
    // function. The EMPTY slice is what the primary constructor builds when told
    // nothing — `AnalysisSlice()` — so the shape carries its own starting value.
    fun withNote(note: AnalysisNote): AnalysisSlice = copy(notes = notes + note)

    fun withPublished(text: String): AnalysisSlice = copy(published = published + text)
}

File: agent-driven-architecture-main/examples/typescript/src/blocks/analysis/slice.ts
// ── blocks/analysis/slice — the tiering block's own state ──────────────────
// PURE, copy-on-write, no I/O, no clock, no Actor.
//
// A note stores the WHOLE sealed `Recall` plus the `at` the fold was given. That
// pairing is deliberate: the age of a recalled conclusion is DERIVED from those
// two values (`ageOf(note.recall, note.at)`), never captured at read time — so
// it is identical on a re-fold instead of drifting with wall-clock time.

import type { Timestamp } from "@adr/spine/pure/ids";
import type { Recall } from "@adr/spine/pure/staged";

export interface AnalysisNote {
  readonly at: Timestamp;
  readonly recall: Recall;
}

export interface AnalysisSlice {
  /** what the FAST tier recalled, in order */
  readonly notes: readonly AnalysisNote[];
  /** what the DEEP tier concluded, in order */
  readonly published: readonly string[];
}

export const emptyAnalysisSlice: AnalysisSlice = { notes: [], published: [] };

export function withNote(slice: AnalysisSlice, note: AnalysisNote): AnalysisSlice {
  return { ...slice, notes: [...slice.notes, note] };
}

export function withPublished(slice: AnalysisSlice, text: string): AnalysisSlice {
  return { ...slice, published: [...slice.published, text] };
}
